### PR TITLE
Fix issue with 'use explicit cast' and compound assignments

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/56141.

### CODE REVIEW

File: src/Analyzers/CSharp/CodeFixes/AddExplicitCast/CSharpAddExplicitCastCodeFixProvider.cs [hunk 1]
```diff
@@ -81,21 +72,39 @@ out ImmutableArray<(ExpressionSyntax, ITypeSymbol)> potentialConversionTypes)
...
                     document, semanticModel, root, targetAttributeArgument, attributeArgumentList, attributeNode, cancellationToken));
             }
         }
+    }
+
+    protected override (SyntaxNode finalTarget, SyntaxNode finalReplacement) Cast(
+        SemanticModel semanticModel, ExpressionSyntax targetNode, ITypeSymbol conversionType)
+    {
+        // The compiler is very ambiguous with assignment expressions `(a += b)`.  An error on it may be an error on the
+        // entire expression or on the RHS of the assignment. Have to reverse engineer what it is doing here.
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/AddExplicitCast/CSharpAddExplicitCastCodeFixProvider.cs:93]: this is the main fix.

File: src/Analyzers/Core/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs [hunk 2]
```diff
@@ -23,18 +22,18 @@ internal abstract partial class AbstractAddExplicitCastCodeFixProvider<TExpressi
...
-    /// Give a set of least specific types with a limit, and the part exceeding the limit doesn't show any code fix,
-    /// but logs telemetry
+    /// Give a set of least specific types with a limit, and the part exceeding the limit doesn't show any code fix, but
+    /// logs telemetry.
     /// </summary>
     private const int MaximumConversionOptions = 3;
 
     protected abstract TExpressionSyntax Cast(TExpressionSyntax expression, ITypeSymbol type);
-    protected abstract void GetPartsOfCastOrConversionExpression(TExpressionSyntax expression, out SyntaxNode type, out SyntaxNode castedExpression);
+    protected abstract void GetPartsOfCastOrConversionExpression(TExpressionSyntax expression, out SyntaxNode type, out TExpressionSyntax castedExpression);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs:31]: stronger type.

File: src/Analyzers/Core/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs [hunk 3]
```diff
@@ -43,17 +42,28 @@ internal abstract partial class AbstractAddExplicitCastCodeFixProvider<TExpressi
...
+    protected ImmutableArray<(TExpressionSyntax node, ITypeSymbol type)> GetPotentialTargetTypes(
         Document document, SemanticModel semanticModel, SyntaxNode root,
-        string diagnosticId, TExpressionSyntax spanNode, CancellationToken cancellationToken,
-        out ImmutableArray<(TExpressionSyntax node, ITypeSymbol type)> potentialConversionTypes);
+        string diagnosticId, TExpressionSyntax spanNode, CancellationToken cancellationToken)
+    {
+        using var _ = ArrayBuilder<(TExpressionSyntax node, ITypeSymbol type)>.GetInstance(out var candidates);
+
+        this.AddPotentialTargetTypes(document, semanticModel, root, diagnosticId, spanNode, candidates, cancellationToken);
+        candidates.RemoveDuplicates();
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs:55]: sharing common code between C# and VB

**Comment by CyrusNajmabadi** [comment]
@akhera99  ptal :)

File: src/Analyzers/Core/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs [hunk 4]
```diff
@@ -67,99 +77,108 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
...
-                    secondConversion is { IsImplicit: false, IsReference: true })
+                var castType = semanticModel.GetTypeInfo(castTypeNode, cancellationToken).Type;
+                if (castType != null)
                 {
-                    return currentRoot.ReplaceNode(
-                        targetNode,
-                        this.Cast((TExpressionSyntax)castedExpression, conversionType)
-                            .WithTriviaFrom(targetNode)
-                            .WithAdditionalAnnotations(Simplifier.Annotation));
+                    var firstConversion = semanticFacts.ClassifyConversion(semanticModel, castedExpression, castType);
```
Comment (ToddGrun) [inline src/Analyzers/Core/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs:146]: nit: could move the def of semanticFacts to here

**Review by ToddGrun — APPROVED** [review]

File: src/Analyzers/Core/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/Core/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs:146]: yup.  though i'm ok with this.

**Pull request merged**